In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_df = pd.read_csv("Resources/lending_data.csv")

# Review the DataFrame
lending_df

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_df['loan_status']

# Separate the X variable, the features
X = lending_df.drop(columns=['loan_status'])

In [4]:
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
y.value_counts()

loan_status
0    75036
1     2500
Name: count, dtype: int64

In [7]:
## 2500 of the loans are at risk for default.

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [8]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split & Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lrm_fit = logistic_regression_model.fit(X_train, y_train)

C:\Users\mberry\anaconda\Anaconda3\envs\dev\lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
testing_prediction = lrm_fit.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [11]:
# Print the balanced_accuracy score of the model
testing_accuracy = balanced_accuracy_score(y_test, testing_prediction)
print(f'The Testing Accuracy with the original data is: {testing_accuracy}')

The Testing Accuracy with the original data is: 0.967989851522121


In [12]:
# Generate a confusion matrix for the model
testing_confusion = confusion_matrix(y_test, testing_prediction)
testing_confusion_df = pd.DataFrame(testing_confusion, 
                                    index = ['Actual Healthy Loans (low-risk)','Actual Non-Healthy Loans (high-risk)'],
                                    columns = ['Predicted Healthy Loans (low-risk)', 'Predicted Non-Healthy Loans (high-risk)'])
testing_confusion_df

Predicted Healthy Loans (low-risk)  \
Actual Healthy Loans (low-risk)                                    18655   
Actual Non-Healthy Loans (high-risk)                                  36   

                                      Predicted Non-Healthy Loans (high-risk)  
Actual Healthy Loans (low-risk)                                           110  
Actual Non-Healthy Loans (high-risk)                                      583

In [13]:
# Print the classification report for the model
testing_report = classification_report(y_test, testing_prediction)
print(f'The Classification Report with the original data is:\n {testing_report}')

The Classification Report with the original data is:
               precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.94      0.89       619

    accuracy                           0.99     19384
   macro avg       0.92      0.97      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** For healthy loans ('0') the percision is 100%, the recall is 99%, and the f1-score is 100%, meaning that the model does really well when perdicting instances. For high-risk loans ('1') the percision is 85%, the recall is 91%, and the f1-score is 88%, meaning that the model does moderately well in predicting instances. It predicts Actual Healthy Loans really well, but 36 loans were "False Positives," meaning they were high risk loans but predicted as healthy.  

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [15]:
#!pip install scikit-learn
#!pip install -U imbalanced-learn

  Obtaining dependency information for imbalanced-learn from https://files.pythonhosted.org/packages/92/e8/86c36e1b13007ca9c89381adac6c078cfc8fb71841a76c08a3fe3eca91d3/imbalanced_learn-0.12.0-py3-none-any.whl.metadata
   ---------------------------------------- 0.0/257.7 kB ? eta -:--:--
   - -------------------------------------- 10.2/257.7 kB ? eta -:--:--
   - -------------------------------------- 10.2/257.7 kB ? eta -:--:--
   ---- ---------------------------------- 30.7/257.7 kB 262.6 kB/s eta 0:00:01
   ------ -------------------------------- 41.0/257.7 kB 245.8 kB/s eta 0:00:01
   ---------- ---------------------------- 71.7/257.7 kB 393.8 kB/s eta 0:00:01
   ---------------------- --------------- 153.6/257.7 kB 654.6 kB/s eta 0:00:01
   --------------------------------- ---- 225.3/257.7 kB 765.3 kB/s eta 0:00:01
   -------------------------------------  256.0/257.7 kB 714.4 kB/s eta 0:00:01
   -------------------------------------- 257.7/257.7 kB 689.2 kB/s eta 0:00:00
  Attem

In [17]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
random_oversampler_model = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = random_oversampler_model.fit_resample(X_train, y_train)

In [18]:
# Count the distinct values of the resampled labels data
y_resampled.value_counts()

loan_status
0    56271
1    56271
Name: count, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [19]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_random_model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
lr_random_model = logistic_regression_random_model.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
random_prediction = lr_random_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [20]:
# Print the balanced_accuracy score of the model
random_accuracy = balanced_accuracy_score(y_test, random_prediction)
print(f'The Testing Accuracy with the resampled data is: {random_accuracy}')

The Testing Accuracy with the resampled data is: 0.9935981855334257


In [21]:
# Generate a confusion matrix for the model
random_confusion = confusion_matrix(y_test, random_prediction)
random_confusion_df = pd.DataFrame(random_confusion, 
                                    index = ['Actual Healthy Loans (low-risk)','Actual Non-Healthy Loans (high-risk)'],
                                    columns = ['Predicted Healthy Loans (low-risk)', 'Predicted Non-Healthy Loans (high-risk)'])
random_confusion_df

Predicted Healthy Loans (low-risk)  \
Actual Healthy Loans (low-risk)                                    18646   
Actual Non-Healthy Loans (high-risk)                                   4   

                                      Predicted Non-Healthy Loans (high-risk)  
Actual Healthy Loans (low-risk)                                           119  
Actual Non-Healthy Loans (high-risk)                                      615

In [23]:
# Print the classification report for the model
random_report = classification_report(y_test, random_prediction)
print(f'The Classification Report with the resampled data is:\n {random_report}')

The Classification Report with the resampled data is:
               precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The oversampled data generated an accuracy score of 99% which turned out to be slightly higher than the original model (97%). The oversampled model performs better since it limited labeling the amount of non-healthy (high-risk) loans as healthy (low-risk). A lending company would want to go with the oversampled model since it would be more costly for a lending company to identify non-healthy loans as healthy due to loss of funds, where as healthy loans being identified as non-healthy is not ideal but safer due to no loss in terms of money. 